File: SystemDesign/concepts/system-design/encryption/python/02_asymmetric_rsa.py
"""
ASYMMETRIC ENCRYPTION — RSA in Python

WHY "ASYMMETRIC"?
  Two DIFFERENT keys — one encrypts, the other decrypts.
  Public key  = share with everyone (anyone can encrypt for you)
  Private key = keep secret       (only YOU can decrypt)

RSA:
  Named after Rivest, Shamir, Adleman (1977)
  Security based on: factoring huge numbers is computationally hard
  Used for: key exchange, digital signatures, certificates

Install: pip install cryptography
"""

from cryptography.hazmat.primitives.asymmetric import rsa, padding
from cryptography.hazmat.primitives import hashes, serialization

# ── 1. Generate RSA Key Pair ──────────────────────────────────────────────────
print("=== 1. Generate RSA-2048 Key Pair ===")

private_key = rsa.generate_private_key(
    public_exponent=65537,   # standard value, always use this
    key_size=2048            # 2048-bit = secure, 4096 = extra secure
)
public_key = private_key.public_key()

# Export keys as PEM strings (what you'd store in a .pem file)
private_pem = private_key.private_bytes(
    encoding=serialization.Encoding.PEM,
    format=serialization.PrivateFormat.PKCS8,
    encryption_algorithm=serialization.NoEncryption()
)
public_pem = public_key.public_bytes(
    encoding=serialization.Encoding.PEM,
    format=serialization.PublicFormat.SubjectPublicKeyInfo
)

print("Private key (first 60 chars):", private_pem.decode()[:60])
print("Public  key (first 60 chars):", public_pem.decode()[:60])

# ── 2. Encrypt with PUBLIC key → Decrypt with PRIVATE key ────────────────────
print("\n=== 2. RSA Encrypt/Decrypt ===")

message = b"Secret message from Alice to Bob"
print("Original:", message.decode())

# OAEP = Optimal Asymmetric Encryption Padding (use this, never raw RSA)
ciphertext = public_key.encrypt(
    message,
    padding.OAEP(
        mgf=padding.MGF1(algorithm=hashes.SHA256()),
        algorithm=hashes.SHA256(),
        label=None
    )
)
print("Encrypted (hex):", ciphertext.hex()[:50] + "...")
print("Ciphertext length:", len(ciphertext), "bytes")  # always 256 bytes for 2048-bit RSA

# Only private key can decrypt
decrypted = private_key.decrypt(
    ciphertext,
    padding.OAEP(
        mgf=padding.MGF1(algorithm=hashes.SHA256()),
        algorithm=hashes.SHA256(),
        label=None
    )
)
print("Decrypted:", decrypted.decode())
print("Match?", decrypted == message)

# ── 3. Digital Signature — Sign with PRIVATE, Verify with PUBLIC ─────────────
print("\n=== 3. Digital Signature ===")
# Signature proves: (a) data wasn't tampered, (b) sender owns the private key

document = b"Transfer $1000 to Alice signed by Bob"

# Bob signs with his PRIVATE key
signature = private_key.sign(
    document,
    padding.PSS(
        mgf=padding.MGF1(hashes.SHA256()),
        salt_length=padding.PSS.MAX_LENGTH
    ),
    hashes.SHA256()
)
print("Signature (hex):", signature.hex()[:50] + "...")

# Anyone with Bob's PUBLIC key can verify
try:
    public_key.verify(
        signature,
        document,
        padding.PSS(
            mgf=padding.MGF1(hashes.SHA256()),
            salt_length=padding.PSS.MAX_LENGTH
        ),
        hashes.SHA256()
    )
    print("Signature VALID — document is authentic")
except Exception:
    print("Signature INVALID — document was tampered!")

# Tamper test — modify the document
tampered = b"Transfer $9999 to Eve  signed by Bob"
try:
    public_key.verify(
        signature,
        tampered,
        padding.PSS(
            mgf=padding.MGF1(hashes.SHA256()),
            salt_length=padding.PSS.MAX_LENGTH
        ),
        hashes.SHA256()
    )
    print("Signature VALID")
except Exception:
    print("Tamper detected — signature INVALID (expected!)")

print("\n--- RSA LIMITATION ---")
print("RSA max message size: ~190 bytes for 2048-bit key with OAEP")
print("Can't encrypt large files directly with RSA")
print("Solution: use RSA to share a small AES key, AES encrypts the file")
print("See 03_https_handshake.py and 05_dek_kek.py")

File: SystemDesign/concepts/system-design/encryption/js/02_asymmetric_rsa.js
/**
 * ASYMMETRIC ENCRYPTION — RSA
 *
 * WHY "ASYMMETRIC"?
 *   TWO different keys — one encrypts, the OTHER decrypts.
 *   Not the same key both ways = ASYMMETRIC
 *
 *   Public key  → share with everyone (encrypt with this)
 *   Private key → keep secret (decrypt with this)
 *
 *   encrypt("hello", publicKey)  → "xK9#mP..."
 *   decrypt("xK9#mP...", privateKey) → "hello"
 *   Different keys = ASYMMETRIC
 *
 * WHY PUBLIC/PRIVATE?
 *   Think of a mailbox:
 *   Public key  = the mail slot (anyone can drop mail in)
 *   Private key = the key to open the box (only you can read)
 *
 *   Anyone can encrypt → only YOU can decrypt.
 *   You can publish your public key on your website — safe.
 *
 * RSA (Rivest–Shamir–Adleman, 1977):
 *   Based on the mathematical difficulty of factoring large numbers.
 *   If n = p × q (two huge primes), finding p and q from n alone
 *   is computationally infeasible (would take millions of years).
 *
 *   Key sizes:
 *   1024 bit → broken (don't use)
 *   2048 bit → minimum acceptable today
 *   4096 bit → strong, used for certificates
 *
 * RSA weakness:
 *   SLOW — 1000x slower than AES
 *   Can only encrypt small data (limited by key size)
 *   → Never encrypt bulk data with RSA
 *   → Use RSA to encrypt the AES key only (hybrid encryption)
 */

const crypto = require('crypto');

// ── Generate RSA key pair ─────────────────────────────────────────────────────
console.log('Generating 2048-bit RSA key pair...');

const { publicKey, privateKey } = crypto.generateKeyPairSync('rsa', {
    modulusLength: 2048,                    // key size in bits
    publicKeyEncoding:  { type: 'pkcs1', format: 'pem' },
    privateKeyEncoding: { type: 'pkcs1', format: 'pem' }
});

console.log('\n--- PUBLIC KEY (safe to share) ---');
console.log(publicKey);

console.log('--- PRIVATE KEY (keep secret!) ---');
console.log(privateKey.substring(0, 100) + '...[truncated]');

// ── Encrypt with PUBLIC key ───────────────────────────────────────────────────
const secret = "AES-key-1234567890abcdef12345678";   // 32 bytes — an AES key
console.log('\nOriginal secret:', secret);

const encrypted = crypto.publicEncrypt(
    { key: publicKey, padding: crypto.constants.RSA_PKCS1_OAEP_PADDING },
    Buffer.from(secret)
);
console.log('\nEncrypted (base64):', encrypted.toString('base64').substring(0, 60) + '...');

// ── Decrypt with PRIVATE key ──────────────────────────────────────────────────
const decrypted = crypto.privateDecrypt(
    { key: privateKey, padding: crypto.constants.RSA_PKCS1_OAEP_PADDING },
    encrypted
);
console.log('\nDecrypted:', decrypted.toString());
console.log('Match?', decrypted.toString() === secret);

// ── Digital Signature (reverse use of keys) ───────────────────────────────────
// Sign with PRIVATE key → verify with PUBLIC key
// Proves: "only the owner of the private key sent this"
console.log('\n--- DIGITAL SIGNATURE ---');

const message = "Transfer $1000 to Bob";
const sign = crypto.createSign('SHA256');
sign.update(message);
const signature = sign.sign(privateKey, 'base64');
console.log('Signature:', signature.substring(0, 60) + '...');

// Verify
const verify = crypto.createVerify('SHA256');
verify.update(message);
const isValid = verify.verify(publicKey, signature, 'base64');
console.log('Signature valid?', isValid);  // true

// Tampered message
const verify2 = crypto.createVerify('SHA256');
verify2.update("Transfer $9999 to Bob");   // different message
const isTampered = verify2.verify(publicKey, signature, 'base64');
console.log('Tampered message valid?', isTampered);  // false

console.log('\n--- RSA SUMMARY ---');
console.log('Encrypt:   use PUBLIC key  → only private key holder can decrypt');
console.log('Sign:      use PRIVATE key → anyone with public key can verify');
console.log('Weakness:  SLOW, small data only → use RSA to share AES key, not data');
